tests/release_tests/test_launch/test_kubernetes.py

```diff
@@ -0,0 +1,187 @@
+                                            "secretKeyRef": {
+                                                "name": "aws-session-token",
+                                                "key": "password",
+                                            }
+                                        },
+                                    },
+                                ]
+                            }
+                        ]
+                    }
+                }
+            },
+        }
+    },
+}
+
+
+def test_kubernetes_agent_on_local_process():
+    entity = "launch-release-testing"
+    project = "release-testing"
+    queue = "kubernetes-queue"
+    job_name = "job-source-release-testing-train.py:v1"
+
+    try:
+        # Start launch agent
+        agent_process = run_cmd_async(f"wandb launch-agent -q {queue} -e {entity}")
+
+        # Start run
+        queued_run = launch_add(
+            job=f"{entity}/{project}/{job_name}", queue_name=queue, entity=entity
+        )
+
+        # Assert successful
+        run = queued_run.wait_until_finished()
+        assert run.state == "finished"
+    finally:
+        agent_process.kill()
+        run_cmd("rm -r artifacts")
+
+
+@pytest.mark.timeout(180)
+def test_kubernetes_agent_in_cluster(api_key, base_url):
+    entity = "launch-release-testing"
+    project = "release-testing"
+    queue = "kubernetes-queue"
+    job_name = (
+        "job-305054156030.dkr.ecr.us-east-2.amazonaws.com_release-testing_latest:v0"
+    )
+
+    # get user's wandb API key and AWS creds
+    if not api_key:
+        api_key = get_wandb_api_key(base_url)
+    aws_id = os.getenv("AWS_ACCESS_KEY_ID")
+    aws_secret = os.getenv("AWS_SECRET_ACCESS_KEY")
+    aws_token = os.getenv("AWS_SESSION_TOKEN")
+    _create_wandb_and_aws_secrets(api_key, aws_id, aws_secret, aws_token)
+
+    run_cmd("kubectl apply -f tests/release_tests/test_launch/launch-config.yml")
+    run_cmd("kubectl apply -f tests/release_tests/test_launch/launch-agent.yml")
+
+    # Capture sigint so cleanup occurs even on ctrl-C
+    sigint = signal.getsignal(signal.SIGINT)
+
+    def cleanup(signum, frame):
+        signal.signal(signal.SIGINT, sigint)
+        _cleanup_deployment()
+        sys.exit(1)
+
+    signal.signal(signal.SIGINT, cleanup)
+
+    pod_name = None
+    try:
+        # Start run
+        launch_add(
+            job=f"{entity}/{project}/{job_name}",
+            queue_name=queue,
+            entity=entity,
+            config=CONFIG,
+        )
+
+        # Wait to finish. Can't use W&B wait_until_finished() because it's an image-based job
+        pod_name, status = _wait_for_job_completion(entity, project)
+
+        assert (
+            status == "Succeeded"
+        ), "Kubernetes job didn't succeed. Check Kubernetes pods and Docker container output."
+    finally:
+        # Cleanup
+        _cleanup_deployment(pod_name)
+
+
+def _cleanup_deployment(pod_name: Optional[str] = None):
```

Comment:
can we list non-cleaned up pods, and then try to delete from there? I feel like putting the responsibility on the caller to include a pod-name is less ideal. if its tricky then probably not essential, but would be a nice feature of a function called "cleanup_deployment"

Reply:
also thinking about more complex workflows, like a job that queues other jobs (for example a dag-creation job, or a sweep scheduler), not sure we will know every pod name